content/post/2025-02-01-lazy/index.Rmd

```diff
@@ -0,0 +1,132 @@
+---
+
+```{r setup, include=FALSE}
+
+knitr::opts_chunk$set(fig.path = "", comment = "")
+# knitr hook to make images output use Hugo options
+knitr::knit_hooks$set(
+  plot = function(x, options) {
+    hugoopts <- options$hugoopts
+    paste0(
+      "{{<figure src=",
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+
+```
+
+```{r, include = FALSE, cache = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>",
+  out.width = "100%",
+  asciicast_theme = "pkgdown"
+)
+```
+
+In the programming world, laziness can often be a good thing: it is both a human quality that can motivate automation efforts, and a programming concept that avoids wasting resources such as memory.
+Now, when reading code or documentation, seeing the word "lazy" can be confusing, because of its polisemy: it carries several meanings.
+In this post, we will enumerate the different possible definitions of "lazy" in R code.
+
+## Lazy as in lazy evaluation
+
+You might know that R provides lazy evaluation: the arguments of a function are only evaluated if they are accessed.
+In short, you can pass anything as an argument value to a function without any problem as long as the function does not use that value.
+The contrary of lazy evaluation is *eager evaluation*.
+
+The [Advanced R book by Hadley Wickham](https://adv-r.hadley.nz/functions.html#lazy-evaluation) features a very clear introduction to lazy evaluation.
+
+Note that the workhorse of lazy evaluation in base R is a thing called a promise that contains an expression (the recipe for getting a value), an environment (the ingredients that are around), a value.
+The latter is only computed when accessed, and cached once computed.
+
+### What about {future}?
+
+Maybe you have heard of the [future package](https://future.futureverse.org/index.html) by Henrik Bengtsson.
+It provides an implementation in R of futures, a programming concept.
+Its homepage state "In programming, a future is an abstraction for a value that may be available at some point in the future. The state of a future can either be unresolved or resolved."
+
+With futures, you create a future, that is associated to a promise, not the same as the aforementioned base R promise, which is a placeholder for a value and then the value itself.
+The value can be computed asynchronously, which means in parallel.
+Therefore, the futures package allows R programmers to take full advantage of their local computing resources: cores, clusters, etc.
+
+Now, to add confusion beyond the different meaning of "promise" in this context, by default a future is **not lazy**, it is **eager**.
+This means that it is computed immediately unless you specify an alternative method (a "plan").
+Therefore, the use of futures and the use of lazy evaluation are orthogonal concepts.
```

Comment:
orthogonal is a great description, but I think also quite complicated as a word, maybe elaborate what it means in this context